### ISSUE TITLE
Make zoom levels 2/3 more tightly integrated

### ISSUE THREAD

**Issue by Robinlovelace** [cyipt/actdev-ui#15]
Split-out from #14:

Suggestions for slightly tightening the zoom level 2-3 connection:

 
- [ ]  mention A/B Street more prominently in the app, e.g. in the `Study area` UI element (see image below)
- [ ]  allow the user to set the scenario (and possibly other things like time) from the 2nd zoom level
 
 
 ![image](https://user-images.githubusercontent.com/1825120/108263407-25349680-715e-11eb-8726-e0e0bb4284a6.png)

Could that be switched to `Study Area / A/B Street launcher` or similar?

(Updated issue based on comments above.)

**Comment by dabreegster** [comment]
https://github.com/cyipt/actdev/issues/66 has some other open design questions around the UI elements in A/B Street. In particular, I'm not sure how many choices to expose to the user for what traffic is being simulated -- just the base/dutch scenarios, mixing in other guesses for traffic not involving the new site, and letting the user spawn their own trips.

If there's already going to be some way for the user to toggle between looking at a baseline/dutch scenario in zoom level 2, then we could pass a flag into level 3 and start with the appropriate scenario. But if that's not already going to be a prominent UI choice in level 2, I don't think introducing the complexity is worth it.

**Comment by Robinlovelace** [comment]
Good points @dabreegster. I'm more thinking about how we can cross reference the two more. Could be done more effectively in comms/docs/'marketing' but think at least referring to A/B Street more in the main zoom 2 app, e.g. by mentioning it in the UI element that exposes it, is a good idea. So it's more about comms/signposting than actual integration (altough I do think an option to select a scenario, with which the user should already be familiar) would be really cool.

**Comment by mvl22** [comment]
This will get resolved in #13, which is about a proper design of the site-level UI.

**Comment by Siequnu** [comment]
I think there are two issues at play — the mechanics of the integration of the abs view, and the styling thereof to match our design. 

@dabreegster is every UI element on your stylable with a CSS file — and if so can you point me to it?

**Comment by dabreegster** [comment]
> I think there are two issues at play

Agreed. I think the important thing to figure out here first is design; implementation can follow that.

> is every UI element on your stylable with a CSS file

Unfortunately not; it's a custom UI library written in Rust. That top-center panel comes from https://github.com/a-b-street/abstreet/blob/6673521ee9b5318c7c71dbb4a4f05ff8d53a3e36/game/src/sandbox/gameplay/blog.rs#L105. There are themes / style presets for all of the elements. If we want them to match the level 2 UI better, we can.

abst's UX designer uses Figma, here's the "design library" page defining most of these templates: https://www.figma.com/file/RX2nOdY2uPaqGGvWPyDvV6/A-B-Street?node-id=1558%3A7513

If you have any suggested changes, we can figure out how to collaborate best.

**Comment by Siequnu** [comment]
Here is our proposed concept for switching between the map and A/B view. 

The idea is to have a segmented control on the map, which changes between map-view and simulation view.

Having a map-level control is advantageous over both external links (breaking the flow of the website) and having a modal full-screen popup, which feels relatively less integrated.

In order to avoid the user seeing the A/B loading screen, we propose to pre-load  the A/B Street view behind the scenes, so that the switching is then immediate.

<img width="1188" alt="Modern dashboard infographic AB" src="https://user-images.githubusercontent.com/11821606/108627485-da35be80-7490-11eb-8b9b-37190f3b9257.png">
<img width="1068" alt="1" src="https://user-images.githubusercontent.com/11821606/108627484-d86bfb00-7490-11eb-91ce-ec319f11fcfd.png">

**Comment by dabreegster** [comment]
Thanks for the concepts -- they look awesome!

## Design

My primary design concern is that A/B Street hasn't been designed for half-width screens, so the main panels smush together:
![Screenshot from 2021-02-21 09-42-29](https://user-images.githubusercontent.com/1664407/108633448-23eabe80-7429-11eb-9fae-9d9115892a10.png)
I could imagine reorienting some of the controls to be more vertical.

There's an additional info panel for map objects and agents that's less clear how to fit in reduced width:
![Screenshot from 2021-02-21 09-45-07](https://user-images.githubusercontent.com/1664407/108633536-965b9e80-7429-11eb-95a6-9c2758245ed2.png)

And some tables that degrade a fair bit:
![Screenshot from 2021-02-21 09-46-34](https://user-images.githubusercontent.com/1664407/108633561-bc813e80-7429-11eb-9766-67804f40293c.png)

If we decide on the 2nd or 3rd option from https://github.com/cyipt/actdev-ui/issues/13#issuecomment-782863219, more horizontal real estate is retained, and these problems aren't as bad.

If we go with this plan, it'd be helpful to come to a decision sooner about which layout to use, to know how much width we have to design around. I can then try to adapt some of the controls. Yuwen, our UX designer, is busy with her capstone project right now and probably can't help. If you have any spare cycles to help reimagine the controls in this split-screen view, it'd be much appreciated.

## Technical

Pre-loading is a really clever idea; I love it. I'm not sure if there's a way to hint to browsers how to prioritize network requests, but we should make sure the initial .wasm download doesn't affect the initial load time of the page too much. In terms of doing the preload, https://github.com/a-b-street/abstreet/blob/master/game/index.html is how we load and start the .wasm today, but it's doing some tricks just to get a loading screen hooked up. The minimal setup is just https://github.com/a-b-street/abstreet/blob/035b8ee96f3300c6683b14e5f2de6afbc8d7cfef/game/index.html.

@michaelkirk happens to currently be experimenting with embedding A/B Street in a smaller HTML canvas, instead of our current approach, which just overwrites onto the DOM.

We'd have to figure out how to handle switching maps. The brute-force approach would be to delete the canvas and stop the wasm script, then recreate it and pass in a parameter for the new map. I'd hope that browsers would cache the .wasm, but we'd pay the cost of having the browser run the .wasm again. The smarter approach would be to just have abst running in the background once and somehow detect that the map has changed, and trigger switching on its end. I'm not too familiar with wasm/js interactions. A hacky approach would be to change the URL to include the site name as a query param, and make abst constantly check for changed URLs. Likely the proper way is to make abst query some DOM element and get the current site name from that, compare to the loaded map, and switch over if needed.

Likewise, to embed the "Map / Simulation" toggle in abst, I can mimic that button style. Upon pressing "Map", we should be able to find some way of calling a JS function, or even doing something like finding the canvas element with abst and disabling its visibility.

**Comment by Robinlovelace** [comment]
My first thought on this was 'sounds like a lot of work A/B Street side'. Given there are other important dev issues to work on on the A/B Street side, that could benefit the development of the project long term and not just the version of A/B Street that works for the purposes of this project, I would suggest we hold off requesting these changes as a matter of priority given the available time on the project. Having used A/B Street a fair bit, I find it has its own immersive feel and UI elements and require as much screen space as possible and wouldn't want to take away from that from a user perspective (although I thinnk the technical issues outlined above may be more important).

**Comment by dabreegster** [comment]
Prioritizing and being realistic about the workshop deadline, I'd say first round, we should first get the app prefetched in the background, loading maps as we change sites, and taking over the whole screen as a modal. Then time depending, I can try adjusting parts of the UI to be more narrow. But I'll need design help with this second part.

**Comment by Robinlovelace** [comment]
> Prioritizing and being realistic about the workshop deadline, I'd say first round, we should first get the app prefetched in the background, loading maps as we change sites, and taking over the whole screen as a modal.

Sounds good to me. Is pre-fetching the basemap going to take time and add complexity though? Thinking of the user with low bandwidth who is taking a look at multiple sites in quick succession. I don't think the load times are too slow so I'm skeptical about the idea of 'prefetching', although I'm not 100% sure what that would look like technically and for what impact it would have on the majority of users who just want to take a quick look.

My current thinking considering the ideas above is that 'loose integration' whereby we signpost to the A/B Side subtly is a good option from user, technical and pragmatic perspectives.

**Comment by dabreegster** [comment]
> Is pre-fetching the basemap going to take time and add complexity though?

That's a good concern; loading the different maps in abst if the user never even clicks through is a waste. The loading time for switching maps is also much less than the time to initially boot the wasm app. So we can treat this as an experiment and just start with preloading in the background.

**Comment by dabreegster** [comment]
I'll try the prefetching tomorrow. But it'd also be good to get feedback from @mvl22 and @Siequnu on the width issue, so I can prioritize that or not. Thanks!

**Comment by Siequnu** [comment]
> I'll try the prefetching tomorrow. But it'd also be good to get feedback from @mvl22 and @Siequnu on the width issue, so I can prioritize that or not. Thanks!

I agree with @Robinlovelace, and would place the width issue further down the priority list. As a first milestone, let's keep the integration to a full-screen modal or external link for now.

After I have chopped all the wireframes and we are fetching and using dynamic data from the ACTDEV side let's revisit the integration issue.

**Comment by dabreegster** [comment]
No urgency here, just recording some TODOs for myself.

I realized another issue with tight integration: abst constantly updates the URL, so you can copy it and immediately warp to the same map and viewport. If we prefetch and toggle, right now that URL'd go straight to the page just with the WASM app.

Some next steps for background prefetching:
- Change maps only once the user clicks "simulation". The idea Martin and I came up with is to have abst attach a callback to a hidden form field or something that has the site name. When "simulation" is clicked, the JS can populate that field with the site, and abst will load the right map.
- I need to verify abst doesn't do any work while it's hidden. I'm not sure if the `winit` library delivers events for minimized windows, or hidden web canvases.

**Comment by mvl22** [comment]
Personally I think that if the user is not on a mobile, we should set a timer on first load and start pre-loading after about 10 seconds in the background, speculatively in case the simulation button is clicked on.

**Comment by Siequnu** [comment]
We have reached the integration for the MVP so I'll close this. We can revisit in the future if necessary.
